prepare/metrics/rag.py

```diff
@@ -174,4 +176,120 @@
+            CopyFields(
+                field_to_field=[("ground_truths_context_ids", "references")],
+                use_query=True,
+            ),
+        ],
+        metric=f"metrics.{metric_name}",
+    )
+    add_to_catalog(metric, f"metrics.rag.{metric_name}", overwrite=True)
+
+
+context_relevance = MetricPipeline(
+    main_score="score",
+    preprocess_steps=[
+        CopyFields(field_to_field=[("contexts", "references")], use_query=True),
+        CopyFields(
+            field_to_field=[("question", "prediction")],
+            use_query=True,
+        ),
+    ],
+    metric="metrics.perplexity_q.flan_t5_small",
+)
+add_to_catalog(context_relevance, "metrics.rag.context_relevance", overwrite=True)
+
+for new_catalog_name, base_catalog_name in [
+    ("metrics.rag.faithfulness", "metrics.token_overlap"),
+    ("metrics.rag.k_precision", "metrics.token_overlap"),
+    ("metrics.rag.bert_k_precision", "metrics.bert_score.deberta_xlarge_mnli"),
+]:
+    metric = MetricPipeline(
+        main_score="score",
+        preprocess_steps=[
+            CopyFields(field_to_field=[("contexts", "references")], use_query=True),
+            CopyFields(
+                field_to_field=[("answer", "prediction")],
+                use_query=True,
+            ),
+        ],
+        metric=base_catalog_name,
+        postpreprocess_steps=[
+            CopyFields(
+                field_to_field=[
+                    ("score/instance/precision", "score/instance/score"),
+                ],
+                use_query=True,
+            )
+        ],
+    )
+    add_to_catalog(metric, new_catalog_name, overwrite=True)
+
+for new_catalog_name, base_catalog_name in [
+    ("metrics.rag.answer_correctness", "metrics.token_overlap"),
+    ("metrics.rag.recall", "metrics.token_overlap"),
+    ("metrics.rag.bert_recall", "metrics.bert_score.deberta_xlarge_mnli"),
+]:
+    metric = MetricPipeline(
+        main_score="score",
+        preprocess_steps=[
+            CopyFields(
+                field_to_field=[("ground_truths", "references")], use_query=True
+            ),
+            CopyFields(
+                field_to_field=[("answer", "prediction")],
+                use_query=True,
+            ),
+        ],
+        metric=base_catalog_name,
+        postpreprocess_steps=[
+            CopyFields(
+                field_to_field=[
+                    ("score/instance/recall", "score/instance/score"),
+                ],
+                use_query=True,
+            )
+        ],
+    )
+    add_to_catalog(metric, new_catalog_name, overwrite=True)
+
+answer_relevance = MetricPipeline(
+    main_score="score",
+    preprocess_steps=[
+        CopyFields(field_to_field=[("question", "references")], use_query=True),
+        CopyFields(
+            field_to_field=[("answer", "prediction")],
+            use_query=True,
+        ),
+        ListFieldValues(fields=["references"], to_field="references"),
```

Comment:
why is this needed?

Reply:
because this metric compares the answer (as the prediction) to the question (as the reference), so we have to wrap the question by a list, since otherwise it will be a string, and references are expected to be lists

Reply:
got it, i will add this explanation as a comment in the code.. tks!